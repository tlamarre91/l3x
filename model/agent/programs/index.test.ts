import { expect, it } from "vitest";
import * as programs from ".";

it("parse works", () => {
  const parsed = programs.parse(PARSEABLE_PROGRAM);
  expect(parsed).toMatchSnapshot();
});

it("parse throws", () => {
  const tryParse = () => programs.parse(UNPARSEABLE_PROGRAM);
  expect(tryParse).toThrowErrorMatchingSnapshot();
});

it("parse+compile works", () => {
  const parsed = programs.parse(COMPILEABLE_PROGRAM_WITH_SPACES);
  const compiled = programs.compile(parsed);
  expect(compiled).toMatchSnapshot();
});

it("parse+compile throws", () => {
  const parsed = programs.parse(UNCOMPILEABLE_PROGRAM);
  const tryCompile = () => {
    programs.compile(parsed);
  };
  expect(tryCompile).toThrowErrorMatchingSnapshot();
});

const PARSEABLE_PROGRAM = `def start
echo hello world
test $peeklast > 0

def loopdeloop
echo again 1 $loop
echo again 2 $loop
go loopdeloop

def moveover
go f
go start
`;

const UNPARSEABLE_PROGRAM = `def start
echo hello world
scan file 0 20
test $peeklast > 0
branch loopdeloop moveover

def loopdeloop
echo again 1 $loop
echo again 2 $loop
go loopdeloop

def moveover
cross f
go start
`;

const COMPILEABLE_PROGRAM_WITH_SPACES = `   def start
        echo hey dudes
              echo this is about it huh?

def end
echo bye dudes
`;

const UNCOMPILEABLE_PROGRAM = `echo hey dudes
echo should've started with def

def end
echo bye dudes
`;

const SANDBOX_PROGRAM = `def start
echo hello sandbox
`;
