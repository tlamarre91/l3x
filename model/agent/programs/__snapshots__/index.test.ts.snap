// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse throws 1`] = `[Error: first token must be keyword; got scan]`;

exports[`parse works 1`] = `
{
  "codeLines": [
    "def start",
    "echo hello world",
    "test $peeklast > 0",
    "",
    "def loopdeloop",
    "echo again 1 $loop",
    "echo again 2 $loop",
    "go loopdeloop",
    "",
    "def moveover",
    "go f",
    "go start",
    "",
  ],
  "statements": [
    {
      "start": {
        "column": 0,
        "line": 0,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 0,
          },
          "symbol": "def",
        },
        {
          "start": {
            "column": 4,
            "line": 0,
          },
          "symbol": "start",
        },
      ],
    },
    {
      "start": {
        "column": 0,
        "line": 1,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 1,
          },
          "symbol": "echo",
        },
        {
          "start": {
            "column": 5,
            "line": 1,
          },
          "symbol": "hello",
        },
        {
          "start": {
            "column": 11,
            "line": 1,
          },
          "symbol": "world",
        },
      ],
    },
    {
      "start": {
        "column": 0,
        "line": 2,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 2,
          },
          "symbol": "test",
        },
        {
          "start": {
            "column": 5,
            "line": 2,
          },
          "symbol": "$peeklast",
        },
        {
          "start": {
            "column": 15,
            "line": 2,
          },
          "symbol": ">",
        },
        {
          "start": {
            "column": 17,
            "line": 2,
          },
          "symbol": "0",
        },
      ],
      "type": "test",
    },
    {
      "start": {
        "column": 0,
        "line": 4,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 4,
          },
          "symbol": "def",
        },
        {
          "start": {
            "column": 4,
            "line": 4,
          },
          "symbol": "loopdeloop",
        },
      ],
    },
    {
      "start": {
        "column": 0,
        "line": 5,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 5,
          },
          "symbol": "echo",
        },
        {
          "start": {
            "column": 5,
            "line": 5,
          },
          "symbol": "again",
        },
        {
          "start": {
            "column": 11,
            "line": 5,
          },
          "symbol": "1",
        },
        {
          "start": {
            "column": 13,
            "line": 5,
          },
          "symbol": "$loop",
        },
      ],
    },
    {
      "start": {
        "column": 0,
        "line": 6,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 6,
          },
          "symbol": "echo",
        },
        {
          "start": {
            "column": 5,
            "line": 6,
          },
          "symbol": "again",
        },
        {
          "start": {
            "column": 11,
            "line": 6,
          },
          "symbol": "2",
        },
        {
          "start": {
            "column": 13,
            "line": 6,
          },
          "symbol": "$loop",
        },
      ],
    },
    {
      "start": {
        "column": 0,
        "line": 7,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 7,
          },
          "symbol": "go",
        },
        {
          "start": {
            "column": 3,
            "line": 7,
          },
          "symbol": "loopdeloop",
        },
      ],
    },
    {
      "start": {
        "column": 0,
        "line": 9,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 9,
          },
          "symbol": "def",
        },
        {
          "start": {
            "column": 4,
            "line": 9,
          },
          "symbol": "moveover",
        },
      ],
    },
    {
      "start": {
        "column": 0,
        "line": 10,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 10,
          },
          "symbol": "go",
        },
        {
          "start": {
            "column": 3,
            "line": 10,
          },
          "symbol": "f",
        },
      ],
    },
    {
      "start": {
        "column": 0,
        "line": 11,
      },
      "tokens": [
        {
          "start": {
            "column": 0,
            "line": 11,
          },
          "symbol": "go",
        },
        {
          "start": {
            "column": 3,
            "line": 11,
          },
          "symbol": "start",
        },
      ],
    },
  ],
}
`;

exports[`parse+compile throws 1`] = `[Error: statement is not part of a state definition]`;

exports[`parse+compile with test command 1`] = `
{
  "procedures": Map {
    "start" => {
      "commands": [
        {
          "comparison": {
            "comparison": "=",
            "type": "comparison",
          },
          "instruction": "test",
          "leftOperand": {
            "register": "$pc",
            "type": "ref",
          },
          "output": {
            "register": "$c",
            "type": "ref",
          },
          "rightOperand": {
            "type": "literal",
            "value": "0",
          },
        },
      ],
      "name": "start",
    },
  },
  "program": {
    "codeLines": [
      "",
      "def start",
      "test $pc = 0",
      "",
    ],
    "statements": [
      {
        "start": {
          "column": 0,
          "line": 1,
        },
        "tokens": [
          {
            "start": {
              "column": 0,
              "line": 1,
            },
            "symbol": "def",
          },
          {
            "start": {
              "column": 4,
              "line": 1,
            },
            "symbol": "start",
          },
        ],
      },
      {
        "start": {
          "column": 0,
          "line": 2,
        },
        "tokens": [
          {
            "start": {
              "column": 0,
              "line": 2,
            },
            "symbol": "test",
          },
          {
            "start": {
              "column": 5,
              "line": 2,
            },
            "symbol": "$pc",
          },
          {
            "start": {
              "column": 9,
              "line": 2,
            },
            "symbol": "=",
          },
          {
            "start": {
              "column": 11,
              "line": 2,
            },
            "symbol": "0",
          },
        ],
        "type": "test",
      },
    ],
  },
  "sourceMap": Map {
    {
      "commands": [
        {
          "comparison": {
            "comparison": "=",
            "type": "comparison",
          },
          "instruction": "test",
          "leftOperand": {
            "register": "$pc",
            "type": "ref",
          },
          "output": {
            "register": "$c",
            "type": "ref",
          },
          "rightOperand": {
            "type": "literal",
            "value": "0",
          },
        },
      ],
      "name": "start",
    } => {
      "column": 0,
      "line": 1,
    },
    {
      "register": "$pc",
      "type": "ref",
    } => {
      "column": 5,
      "line": 2,
    },
    {
      "comparison": "=",
      "type": "comparison",
    } => {
      "column": 9,
      "line": 2,
    },
    {
      "type": "literal",
      "value": "0",
    } => {
      "column": 11,
      "line": 2,
    },
    {
      "comparison": {
        "comparison": "=",
        "type": "comparison",
      },
      "instruction": "test",
      "leftOperand": {
        "register": "$pc",
        "type": "ref",
      },
      "output": {
        "register": "$c",
        "type": "ref",
      },
      "rightOperand": {
        "type": "literal",
        "value": "0",
      },
    } => {
      "column": 0,
      "line": 2,
    },
  },
}
`;

exports[`parse+compile works 1`] = `
{
  "procedures": Map {
    "start" => {
      "commands": [
        {
          "instruction": "echo",
          "operands": [
            {
              "type": "literal",
              "value": "hey",
            },
            {
              "type": "literal",
              "value": "dudes",
            },
          ],
        },
        {
          "instruction": "echo",
          "operands": [
            {
              "type": "literal",
              "value": "this",
            },
            {
              "type": "literal",
              "value": "is",
            },
            {
              "type": "literal",
              "value": "about",
            },
            {
              "type": "literal",
              "value": "it",
            },
            {
              "type": "literal",
              "value": "huh?",
            },
          ],
        },
      ],
      "name": "start",
    },
    "end" => {
      "commands": [
        {
          "instruction": "echo",
          "operands": [
            {
              "type": "literal",
              "value": "bye",
            },
            {
              "type": "literal",
              "value": "dudes",
            },
          ],
        },
      ],
      "name": "end",
    },
  },
  "program": {
    "codeLines": [
      "   def start",
      "        echo hey dudes",
      "              echo this is about it huh?",
      "",
      "def end",
      "echo bye dudes",
      "",
    ],
    "statements": [
      {
        "start": {
          "column": 3,
          "line": 0,
        },
        "tokens": [
          {
            "start": {
              "column": 3,
              "line": 0,
            },
            "symbol": "def",
          },
          {
            "start": {
              "column": 7,
              "line": 0,
            },
            "symbol": "start",
          },
        ],
      },
      {
        "start": {
          "column": 8,
          "line": 1,
        },
        "tokens": [
          {
            "start": {
              "column": 8,
              "line": 1,
            },
            "symbol": "echo",
          },
          {
            "start": {
              "column": 13,
              "line": 1,
            },
            "symbol": "hey",
          },
          {
            "start": {
              "column": 17,
              "line": 1,
            },
            "symbol": "dudes",
          },
        ],
      },
      {
        "start": {
          "column": 14,
          "line": 2,
        },
        "tokens": [
          {
            "start": {
              "column": 14,
              "line": 2,
            },
            "symbol": "echo",
          },
          {
            "start": {
              "column": 19,
              "line": 2,
            },
            "symbol": "this",
          },
          {
            "start": {
              "column": 24,
              "line": 2,
            },
            "symbol": "is",
          },
          {
            "start": {
              "column": 27,
              "line": 2,
            },
            "symbol": "about",
          },
          {
            "start": {
              "column": 33,
              "line": 2,
            },
            "symbol": "it",
          },
          {
            "start": {
              "column": 36,
              "line": 2,
            },
            "symbol": "huh?",
          },
        ],
      },
      {
        "start": {
          "column": 0,
          "line": 4,
        },
        "tokens": [
          {
            "start": {
              "column": 0,
              "line": 4,
            },
            "symbol": "def",
          },
          {
            "start": {
              "column": 4,
              "line": 4,
            },
            "symbol": "end",
          },
        ],
      },
      {
        "start": {
          "column": 0,
          "line": 5,
        },
        "tokens": [
          {
            "start": {
              "column": 0,
              "line": 5,
            },
            "symbol": "echo",
          },
          {
            "start": {
              "column": 5,
              "line": 5,
            },
            "symbol": "bye",
          },
          {
            "start": {
              "column": 9,
              "line": 5,
            },
            "symbol": "dudes",
          },
        ],
      },
    ],
  },
  "sourceMap": Map {
    {
      "commands": [
        {
          "instruction": "echo",
          "operands": [
            {
              "type": "literal",
              "value": "hey",
            },
            {
              "type": "literal",
              "value": "dudes",
            },
          ],
        },
        {
          "instruction": "echo",
          "operands": [
            {
              "type": "literal",
              "value": "this",
            },
            {
              "type": "literal",
              "value": "is",
            },
            {
              "type": "literal",
              "value": "about",
            },
            {
              "type": "literal",
              "value": "it",
            },
            {
              "type": "literal",
              "value": "huh?",
            },
          ],
        },
      ],
      "name": "start",
    } => {
      "column": 3,
      "line": 0,
    },
    {
      "type": "literal",
      "value": "hey",
    } => {
      "column": 13,
      "line": 1,
    },
    {
      "type": "literal",
      "value": "dudes",
    } => {
      "column": 17,
      "line": 1,
    },
    {
      "instruction": "echo",
      "operands": [
        {
          "type": "literal",
          "value": "hey",
        },
        {
          "type": "literal",
          "value": "dudes",
        },
      ],
    } => {
      "column": 8,
      "line": 1,
    },
    {
      "type": "literal",
      "value": "this",
    } => {
      "column": 19,
      "line": 2,
    },
    {
      "type": "literal",
      "value": "is",
    } => {
      "column": 24,
      "line": 2,
    },
    {
      "type": "literal",
      "value": "about",
    } => {
      "column": 27,
      "line": 2,
    },
    {
      "type": "literal",
      "value": "it",
    } => {
      "column": 33,
      "line": 2,
    },
    {
      "type": "literal",
      "value": "huh?",
    } => {
      "column": 36,
      "line": 2,
    },
    {
      "instruction": "echo",
      "operands": [
        {
          "type": "literal",
          "value": "this",
        },
        {
          "type": "literal",
          "value": "is",
        },
        {
          "type": "literal",
          "value": "about",
        },
        {
          "type": "literal",
          "value": "it",
        },
        {
          "type": "literal",
          "value": "huh?",
        },
      ],
    } => {
      "column": 14,
      "line": 2,
    },
    {
      "commands": [
        {
          "instruction": "echo",
          "operands": [
            {
              "type": "literal",
              "value": "bye",
            },
            {
              "type": "literal",
              "value": "dudes",
            },
          ],
        },
      ],
      "name": "end",
    } => {
      "column": 0,
      "line": 4,
    },
    {
      "type": "literal",
      "value": "bye",
    } => {
      "column": 5,
      "line": 5,
    },
    {
      "type": "literal",
      "value": "dudes",
    } => {
      "column": 9,
      "line": 5,
    },
    {
      "instruction": "echo",
      "operands": [
        {
          "type": "literal",
          "value": "bye",
        },
        {
          "type": "literal",
          "value": "dudes",
        },
      ],
    } => {
      "column": 0,
      "line": 5,
    },
  },
}
`;

exports[`sandbox 1`] = `
{
  "procedures": Map {
    "start" => {
      "commands": [
        {
          "instruction": "echo",
          "operands": [
            {
              "type": "literal",
              "value": "hey1",
            },
          ],
        },
        {
          "instruction": "echo",
          "operands": [
            {
              "type": "literal",
              "value": "hey2",
            },
          ],
        },
        {
          "instruction": "go",
          "state": {
            "type": "literal",
            "value": "loop",
          },
        },
      ],
      "name": "start",
    },
    "loop" => {
      "commands": [
        {
          "instruction": "echo",
          "operands": [
            {
              "type": "literal",
              "value": "woo1",
            },
          ],
        },
        {
          "instruction": "echo",
          "operands": [
            {
              "type": "literal",
              "value": "woo2",
            },
          ],
        },
        {
          "instruction": "go",
          "state": {
            "type": "literal",
            "value": "start",
          },
        },
      ],
      "name": "loop",
    },
  },
  "program": {
    "codeLines": [
      "def start",
      "echo hey1",
      "echo hey2",
      "go loop",
      "",
      "def loop",
      "echo woo1",
      "echo woo2",
      "go start",
      "",
    ],
    "statements": [
      {
        "start": {
          "column": 0,
          "line": 0,
        },
        "tokens": [
          {
            "start": {
              "column": 0,
              "line": 0,
            },
            "symbol": "def",
          },
          {
            "start": {
              "column": 4,
              "line": 0,
            },
            "symbol": "start",
          },
        ],
      },
      {
        "start": {
          "column": 0,
          "line": 1,
        },
        "tokens": [
          {
            "start": {
              "column": 0,
              "line": 1,
            },
            "symbol": "echo",
          },
          {
            "start": {
              "column": 5,
              "line": 1,
            },
            "symbol": "hey1",
          },
        ],
      },
      {
        "start": {
          "column": 0,
          "line": 2,
        },
        "tokens": [
          {
            "start": {
              "column": 0,
              "line": 2,
            },
            "symbol": "echo",
          },
          {
            "start": {
              "column": 5,
              "line": 2,
            },
            "symbol": "hey2",
          },
        ],
      },
      {
        "start": {
          "column": 0,
          "line": 3,
        },
        "tokens": [
          {
            "start": {
              "column": 0,
              "line": 3,
            },
            "symbol": "go",
          },
          {
            "start": {
              "column": 3,
              "line": 3,
            },
            "symbol": "loop",
          },
        ],
      },
      {
        "start": {
          "column": 0,
          "line": 5,
        },
        "tokens": [
          {
            "start": {
              "column": 0,
              "line": 5,
            },
            "symbol": "def",
          },
          {
            "start": {
              "column": 4,
              "line": 5,
            },
            "symbol": "loop",
          },
        ],
      },
      {
        "start": {
          "column": 0,
          "line": 6,
        },
        "tokens": [
          {
            "start": {
              "column": 0,
              "line": 6,
            },
            "symbol": "echo",
          },
          {
            "start": {
              "column": 5,
              "line": 6,
            },
            "symbol": "woo1",
          },
        ],
      },
      {
        "start": {
          "column": 0,
          "line": 7,
        },
        "tokens": [
          {
            "start": {
              "column": 0,
              "line": 7,
            },
            "symbol": "echo",
          },
          {
            "start": {
              "column": 5,
              "line": 7,
            },
            "symbol": "woo2",
          },
        ],
      },
      {
        "start": {
          "column": 0,
          "line": 8,
        },
        "tokens": [
          {
            "start": {
              "column": 0,
              "line": 8,
            },
            "symbol": "go",
          },
          {
            "start": {
              "column": 3,
              "line": 8,
            },
            "symbol": "start",
          },
        ],
      },
    ],
  },
  "sourceMap": Map {
    {
      "commands": [
        {
          "instruction": "echo",
          "operands": [
            {
              "type": "literal",
              "value": "hey1",
            },
          ],
        },
        {
          "instruction": "echo",
          "operands": [
            {
              "type": "literal",
              "value": "hey2",
            },
          ],
        },
        {
          "instruction": "go",
          "state": {
            "type": "literal",
            "value": "loop",
          },
        },
      ],
      "name": "start",
    } => {
      "column": 0,
      "line": 0,
    },
    {
      "type": "literal",
      "value": "hey1",
    } => {
      "column": 5,
      "line": 1,
    },
    {
      "instruction": "echo",
      "operands": [
        {
          "type": "literal",
          "value": "hey1",
        },
      ],
    } => {
      "column": 0,
      "line": 1,
    },
    {
      "type": "literal",
      "value": "hey2",
    } => {
      "column": 5,
      "line": 2,
    },
    {
      "instruction": "echo",
      "operands": [
        {
          "type": "literal",
          "value": "hey2",
        },
      ],
    } => {
      "column": 0,
      "line": 2,
    },
    {
      "type": "literal",
      "value": "loop",
    } => {
      "column": 3,
      "line": 3,
    },
    {
      "instruction": "go",
      "state": {
        "type": "literal",
        "value": "loop",
      },
    } => {
      "column": 0,
      "line": 3,
    },
    {
      "commands": [
        {
          "instruction": "echo",
          "operands": [
            {
              "type": "literal",
              "value": "woo1",
            },
          ],
        },
        {
          "instruction": "echo",
          "operands": [
            {
              "type": "literal",
              "value": "woo2",
            },
          ],
        },
        {
          "instruction": "go",
          "state": {
            "type": "literal",
            "value": "start",
          },
        },
      ],
      "name": "loop",
    } => {
      "column": 0,
      "line": 5,
    },
    {
      "type": "literal",
      "value": "woo1",
    } => {
      "column": 5,
      "line": 6,
    },
    {
      "instruction": "echo",
      "operands": [
        {
          "type": "literal",
          "value": "woo1",
        },
      ],
    } => {
      "column": 0,
      "line": 6,
    },
    {
      "type": "literal",
      "value": "woo2",
    } => {
      "column": 5,
      "line": 7,
    },
    {
      "instruction": "echo",
      "operands": [
        {
          "type": "literal",
          "value": "woo2",
        },
      ],
    } => {
      "column": 0,
      "line": 7,
    },
    {
      "type": "literal",
      "value": "start",
    } => {
      "column": 3,
      "line": 8,
    },
    {
      "instruction": "go",
      "state": {
        "type": "literal",
        "value": "start",
      },
    } => {
      "column": 0,
      "line": 8,
    },
  },
}
`;
